### PR TITLE
zephyr: remove deprecated Kconfigs

### PR DESCRIPTION
**Pull request by ycsin**
These Kconfigs should have been deprecated for more than 2 releases, remove them:

- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSI`
- `CONFIG_COUNTER_RTC_STM32_CLOCK_LSE`
- `CONFIG_EMUL_EEPROM_AT2X`
- `CONFIG_ETH_STM32_HAL_RANDOM_MAC`
- `CONFIG_ETH_STM32_HAL_USER_STATIC_MAC`
- `CONFIG_ETH_STM32_HAL_MAC3`
- `CONFIG_ETH_STM32_HAL_MAC4`
- `CONFIG_ETH_STM32_HAL_MAC5`
- `CONFIG_IEEE802154_2015`
- `CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ_RADIO_TX_RETRIES`
- `CONFIG_CPP_MAIN`
- `CONFIG_CPLUSPLUS`
- `CONFIG_LIB_CPLUSPLUS`
- `CONFIG_EXCEPTIONS`
- `CONFIG_RTTI`
- `CONFIG_SUPPORT_MINIMAL_LIBC`
- `CONFIG_MINIMAL_LIBC_MALLOC`
- `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE`
- `CONFIG_MINIMAL_LIBC_CALLOC`
- `CONFIG_MINIMAL_LIBC_REALLOCARRAY`
- `CONFIG_PICOLIBC_HEAP_SIZE`
- `CONFIG_BT_DEBUG_LOG`
- `CONFIG_LOG_DOMAIN_ID`
- `CONFIG_NET_L2_IEEE802154_ACK_REPLY`
- `CONFIG_SETTINGS_FS`
- `CONFIG_SETTINGS_FS_DIR`
- `CONFIG_SETTINGS_FS_FILE`
- `CONFIG_SETTINGS_FS_MAX_LINES`

Remove related code in .c/.h as well.

### CODE REVIEW

**Review by jhedberg — DISMISSED** [review]

**Review by PavelVPV — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

File: lib/libc/Kconfig
Comment (keith-packard) [inline lib/libc/Kconfig:75]: Until #65434 is merged, I don't think we should be adding malloc to the default minimal libc configuration as that could increase the size of applications which don't use it. That's because the malloc heap initializer will always end up linked into applications because the common C library bits are included under the `--whole-archive` list.

File: lib/libc/common/Kconfig
Comment (keith-packard) [inline lib/libc/common/Kconfig:56]: You can just remove `&& !MINIMAL_LIBC_CALLOC` here. The goal is to disable calloc when using the minimal C library by default.

File: lib/libc/common/Kconfig
Comment (keith-packard) [inline lib/libc/common/Kconfig:65]: Same here -- leave the line, but remove `&& !MINIMAL_LIBC_REALLOC_ARRAY`

**Review by keith-packard — CHANGES_REQUESTED** [review]
Need to fix the semantic changes to the common C library malloc code when using the minimal C library.

File: lib/libc/Kconfig
Comment (cfriedt) [reply lib/libc/Kconfig:75]: I had a change to remove the sys_init from the minimal libc malloc (common, I guess) and use lazy init instead (init on first use), but it was more complicated than that.

Honestly, that should have a bug / issue number attached to it, because it needs to be fixed. In no universe should we be unconditionally linking malloc buffers and code without an explicit call to malloc or something related.

**Review by ghost — DISMISSED** [review]

File: lib/libc/Kconfig
Comment (ycsin) [reply lib/libc/Kconfig:75]: This is moved from:

```
...

if MINIMAL_LIBC

...

config MINIMAL_LIBC_MALLOC
	bool "[DEPRECATED] Minimal libc malloc implementation"
	default y
	imply COMMON_LIBC_MALLOC
	help
	  [DEPRECATED] Use COMMON_LIBC_MALLOC

	  Enable the minimal libc's implementation of malloc, free, and realloc.
	  Disable if you wish to provide your own implementations of these functions.

...

```

which I have deleted in `lib/libc/minimal/Kconfig`

**Review by jhedberg — DISMISSED** [review]

File: lib/libc/Kconfig
Comment (keith-packard) [reply lib/libc/Kconfig:75]: Yup, but this changes the default. Before, you didn't get malloc unless you asked for it. With the proposed change, you get malloc unless you ask to *not* have it. Unless I'm mis-reading the patch?

Comment (ycsin) [reply lib/libc/Kconfig:75]: Currently, enabling `MINIMAL_LIBC` would enable `MINIMAL_LIBC_MALLOC` by default, which would `imply` `COMMON_LIBC_MALLOC`, so malloc will be enabled unless user explicitly disables `MINIMAL_LIBC_MALLOC` and/or `COMMON_LIBC_MALLOC`.

FWIU, this PR removes the intermediary `MINIMAL_LIBC_MALLOC` & relocate the `imply COMMON_LIBC_MALLOC` directly under the `MINIMAL_LIBC`, so `COMMON_LIBC_MALLOC` will be enabled by default, which is the same as before, unless explicitly disabled by the user, but there's one less Kconfig (`MINIMAL_LIBC_MALLOC`) that a user has to care.

Or maybe I missed something?

**Review by nordic-krch — DISMISSED** [review]

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:630]: Please remove the whole `#if DT_INST_NUM_CLOCKS(0) == 1` section.

**Review by erwango — CHANGES_REQUESTED** [review]
Otherwise LGTM, thanks

**Review by jhedberg — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Review by erwango — APPROVED** [review]

File: drivers/counter/Kconfig.stm32_rtc
Comment (erwango) [inline drivers/counter/Kconfig.stm32_rtc:25]: Please remove the while choice

**Review by erwango — CHANGES_REQUESTED** [review]
Sorry for not seeing this earlier.

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Comment by nashif** [comment]
You probably want to do this in multiple commits, remove unrelated Kconfigs all in one commit does work well and will make reviews difficult.

**Comment by nashif** [comment]
this could go into 3.7 IMO

**Review by omkar3141 — DISMISSED** [review]
Mesh changes look non-controversial.

**Comment by ycsin** [comment]
Split the commit into multiple commits, PTAL, thanks.

File: drivers/counter/Kconfig.stm32_rtc
Comment (ycsin) [reply drivers/counter/Kconfig.stm32_rtc]: removed

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (ycsin) [reply drivers/counter/counter_ll_stm32_rtc.c]: removed

File: lib/libc/common/Kconfig
Comment (ycsin) [reply lib/libc/common/Kconfig:65]: updated

File: lib/libc/common/Kconfig
Comment (ycsin) [reply lib/libc/common/Kconfig:56]: updated

**Review by erwango — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

File: lib/libc/Kconfig
Comment (keith-packard) [reply lib/libc/Kconfig:75]: Nope, *I* missed that minimal libc sets `MINIMAL_LIBC_MALLOC` to `default y`. Thanks for your patience, this looks fine.

**Review by keith-packard — APPROVED** [review]
I missed that minimal C library defaults to malloc enabled. My previous comments were mis-informed.

**Review by stephanosio — APPROVED** [review]
libc and cpp changes LGTM.

**Review by nashif — APPROVED** [review]

**Pull request merged**
